### DISCUSSION TITLE
support #idx (meaning indexing from last)

### DISCUSSION THREAD

**Discussion by SleepyRoy** [vlang/v#11679] [Ideas]
I notice many people miss the minus indexing feature in other languages such as Python.
```Python
# Python
arr[-1]  # last element
arr[-2]  # second from last 
arr[-5:]  # five elements from last
```
However, Alex doesn't like this because it's error-prone. Yes indeed sometimes. So I'm thinking about using a symbol other than minus(`-`) to indicate indexing from last. Say we use `#`.
```V
// V
arr[#1]  // last element
arr[#2]  // second from last
arr[#5..]  // five elements from last
```
And there's a bonus. I notice some want last element to be `-0` rather than `-1`. We know this is not a good idea for the old minus(`-`) system because `-0` and `0` are numerically equal. But by applying the new `#` system this could be solved - `0` is different from `#0`.

1. Do you like this idea?
2. If so, do you like `#` to be the symbol?
3. Do you prefer the last element as `#0` or `#1`?

BTW I personally prefer last element to be `#0`. Because when considering backward indexing you can still assume it's the more familiar forward indexing, and finally add a `#`. Very easy. 
```V
// V - I prefer this.
arr[#0]  // last element
arr[#1]  // second from last 
arr[#4..]  // five elements from last 
```

**Comment by sudo-nice** [comment]
>Do you like this idea?

Yes, it seems to be handy.

>If so, do you like `#` to be the symbol?

It's OK (or maybe `!`).

>Do you prefer the last element as `#0` or `#1`?

`#0` (for consistency with forward indexing).

**Comment by JalonSolov** [comment]
For consistency with every other system that let's you reference back from the the end, it would have to be `#1` to mimic the `-1` the others use.  Assuming `#` is used.

There's no reason to confuse people with "Yes, it's -1 over there, but #0 here."... changing both the symbol in front of the digit AND the digit.  Just changing the symbol will be bad enough.

**Comment by sudo-nice** [comment]
Yes, it might be unexpected for the backward indexing to start from `0`, but for a good reason, I think. We often iterate over an array, using code like this:

```v
fn main() {
  arr := [1, 2, 3, 4, 5]
  for i in 0 .. (arr.len - 1) {
    println(arr[i])
  }
}
```
Now, if we want to change it to index backwards, we can just **add one symbol** (in case of zero indexing from both ends):

```v
fn main() {
  arr := [1, 2, 3, 4, 5]
  for i in 0 .. (arr.len - 1) {
    println(arr[#i])
  }
}
```
But, if we want to change it to index backwards, we need to **make tricky changes in two lines**:
```v
fn main() {
  arr := [1, 2, 3, 4, 5]
  for i in 0 .. (arr.len) {
    println(arr[#(i + 1)])
  }
}
```
If we keep zero-based indexing, is that going to break the mode of backward indexing, which many programmers could get accustomed to? Yes.
But I think we should break stereotypes for good reasons, and we have a pretty good one here.

**Comment by SleepyRoy** [comment]
Yes, I also like the last element to be `#0`, which makes forward and backward indexing rules symmetrical. People like easy-to-remember things.
When writing Python, I experienced such (bad) consideration process for many times - Next I need to get the i-th element from last, and what's the index? `-(i-1)`...oh no, incorrect, should be `-i` as this is backward indexing. I think this is because we're so familiar with the starting-from-zero forward indexing system.
